core/src/main/scala/scala/pickling/internal/package.scala

```diff
@@ -24,10 +23,18 @@ package object internal {
 
     }
   }
-  private[this] var currentRuntimeVar = new AtomicReference[spi.PicklingRuntime](initDefaultRuntime)
-  def currentRuntime: spi.PicklingRuntime = currentRuntimeVar.get
-  // Here we inject a new runtime for usage.
-  def replaceRuntime(r: spi.PicklingRuntime): Unit = {
+  private[this] var currentRuntimeVar = new AtomicReference[PicklingRuntime](initDefaultRuntime)
+  def currentRuntime: PicklingRuntime = currentRuntimeVar.get
+
+  /** Replace the old [[PicklingRuntime]] keeping its state.
+    *
+    * Note that we don't do anything with the [[RefRegistry]] because in
+    * future versions we are going to change how cyclic references work.
+    */
+  def replaceRuntime(r: PicklingRuntime): Unit = {
+    if(r.picklers.isLookupEnabled) {
+      currentRuntime.picklers.dumpStateTo(r.picklers)
```

Comment:
The issue with synchronizing this method is you'd need to ALSO synchronize all other access to the pickler registry, whcih slows down general runtime.   We already use thread-safe lookup (in the eventual consistency version of thread safety), so I'd hate to furtehr slow down lookup with a full up synchronize.


There are ways to force things to initialize without forcing implicitly, specifically if you use the "cake" approach for your default picklers.  It's not pretty, but it is workable.

In any case, as i said, I think what you have is fine, we just need to document WHY we think it's fine and where it is not fine :).   As long as it's documented, users may be surprised, but able to cope.  that's better than surprised and have no idea how to resolve the issue.

The best case would be if we had a legitimate solution to the problem.  Given the current architecture, I think we do not, but I'm happy to hear alternative ideas :)